wasm-engines.md

```diff
@@ -0,0 +1,223 @@
+- **wasm-reduce**: Reduce a wasm file to a smaller one that has the same behavior on a given command.
+- **wasm-shell**: A shell that can load and interpret WebAssembly code.
+
+## Using binaryen tools
+
+In this section we show how to use some of the binaryen tools:
+
+### wasm-as
+
+This tool allows to compile WebAssembly Text Format (wast) to
+WebAssembly Binary Format (wasm).  Example: We have a file called
+`contract.wast` containing this wast code:
+
+```wast
+(module
+    (import "ethereum" "storageStore" (func $storageStore (param i32 i32)))
+    (import "ethereum" "getCodeSize"  (func $getCodeSize (result i32)))
+    (memory 1)
+    (export "main" (func $main))
+    (export "memory" (memory 0))
+    (func $main
+      (i32.store (i32.const 32) (call $getCodeSize))
+      (call $storageStore (i32.const 0) (i32.const 32))))
+```
+
+Running `./wasm-as contract.wast` will generate a `contract.wasm` file.
+
+### wasm-dis
+
+You can take `contract.wasm` or any other `.wasm` file generated by another
+compiler and convert it to WebAssembly Text format (`wast`).  Example: running
+`./wasm-dis contract.wasm -o new_contract.wast` will create the following wast
+code:
+
+```wast
+(module
+ (type $0 (func (param i32 i32)))
+ (type $1 (func (result i32)))
+ (type $2 (func))
+ (import "ethereum" "storageStore" (func $fimport$0 (param i32 i32)))
+ (import "ethereum" "getCodeSize" (func $fimport$1 (result i32)))
+ (memory $0 1)
+ (export "main" (func $0))
+ (export "memory" (memory $0))
+ (func $0 (; 2 ;) (type $2)
+  (i32.store
+   (i32.const 32)
+   (call $fimport$1)
+  )
+  (call $fimport$0
+   (i32.const 0)
+   (i32.const 32)
+  )
+ )
+)
+```
+
+Note it is not _exactly_ the same code we wrote in the first place, it is more
+verbose and the variable names have changed as it was generated based on the wasm binary file.
+
+### wasm-shell
+
+This tool allows you to execute wast files.
+
+Example: This WebAssembly program contains two functions, `main` and `sum`.
+`sum` receives two parameters `$a` and `b` and returns the sum of both
+parameters. `main` calls `sum` using `2` and `3` as parameters, then calls a
+`wasm-shell` provided function called `$print` to show the result.
+
+```wast
+(module
+  (import "spectest" "print" (func $print (param i32)))
+  (memory 1)
+  (export "main" (func $main))
+  (export "memory" (memory 0))
+  (func $main
+    (call $print (call $sum (i32.const 2) (i32.const 3))))
+  (func $sum (param $a i32) (param $b i32) (result i32)
+    (return (i32.add (get_local $a) (get_local $b)))))
+```
+
+You can execute this code by calling the command: `./wasm-shell --entry main
+mycode.wast`, and you will get the following result:
+
+```
+BUILDING MODULE [line: 1]
+(i32.const 5)
+```
+
+Note that you cannot currently execute ewasm contracts using `wasm-shell` because the
+`ethereum` namespace is not provided.
+
+# Wabt
+
+## Getting and compiling wabt
+
+Install development tools (`build-essential`) `cmake`, `make` and `clang`:
```

Comment:
Same comments as above:
- Can we remove the `cmake` installation steps, or move them to a global FAQ doc or something?
- Can we link directly to the official wabt build docs, and their list of tools, rather than listing them all here?